GHzDACs/jump_table.py

```diff
@@ -0,0 +1,480 @@
+        with ``util.littleEndian(my_operation.as_int(), 2)``.
+        """
+        raise NotImplementedError()
+
+
+class IDLE(Operation):
+    NAME = "IDLE"
+
+    def __init__(self, cycles):
+        self.cycles = cycles
+
+    def __str__(self):
+        return "%s %d cycles" % (self.NAME, self.cycles)
+
+    def as_int(self):
+        if not (IDLE_MIN_CYCLES <= self.cycles <= IDLE_MAX_CYCLES):
+            raise ValueError("IDLE num cycles must fit in {} bits".format(
+                IDLE_NUM_BITS
+            ))
+        return self.cycles << 1
+
+
+class CHECK(Operation):
+    NAME = "CHECK"
+
+    def __init__(self, which_daisy_bit, bit_state, next_jump_index):
+        self.which_daisy_bit = which_daisy_bit
+        self.jump_index = next_jump_index
+        self.bit_state = bool(bit_state)
+        raise NotImplementedError('CHECK is not yet implemented')
+
+    def as_int(self):
+        # Op code is 001, so shift 4 bits and add 1
+        return (self.jump_index << 8) + (self.which_daisy_bit << 4) + (int(self.bit_state) << 3) + 1
+
+
+class JUMP(Operation):
+    NAME = "JUMP"
+
+    def __init__(self, next_jump_index):
+        self.jump_index = next_jump_index
+
+    def __str__(self):
+        return '\n'.join([self.NAME, "Next jump index: %d" % self.jump_index])
+
+    def as_int(self):
+        # Op code is 1101 = 13
+        return (self.jump_index << 8) + 13
+
+
+class NOP(Operation):
+    NAME = "NOP"
+
+    def __str__(self):
+        return self.NAME
+
+    def as_int(self):
+        return 5
+
+
+class CYCLE(Operation):
+    NAME = "CYCLE"
+
+    def __init__(self, counter, next_jump_index):
+        """ Create a cycle opcode.
+
+        :param int counter: which counter to use (0-indexed)
+        :param int next_jump_index: JT entry to jump to
+        """
+        self.jump_index = next_jump_index
+        self.counter = counter
+
+    def as_int(self):
+        return (self.jump_index << 8) + (self.counter << 4) + 3
+
+    def __str__(self):
+        return "{}: counter={}, JT index={}".format(
+            self.NAME, self.counter, self.jump_index
+        )
+
+
+class END(Operation):
+    NAME = "END"
+
+    def __str__(self):
+        return self.NAME
+
+    def as_int(self):
+        return 7
+
+
+class JumpTable(object):
+    """
+    The entire jump table.
+    
+    This is a very low level wrapper around the actual FPGA data structure
+    
+    TODO: make self.jumps a property with some checking on type and number
+    """
+    PACKET_LEN = 528
+    COUNTER_BITS = 32  # 32 bit register for counters
+    COUNT_MAX = 2**COUNTER_BITS - 1
+    NUM_COUNTERS = 4
+
+    def __init__(self, start_addr=None, jumps=None, counters=None):
+        """
+
+        :param start_addr: start address
+        :param list[JumpEntry] jumps: the jump table entries
+        :param list[int] counters: the counter values
+        """
+        self._startAddr = 0
+        if counters is None:
```

Comment:
This whole business might be well packaged into a method called `_set_up_counters` or whatever.